File: leetcode/Medium/0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        q = deque([0])
        visited = set()
        while q:
            start = q.popleft()
            if start == len(s):
                return True
            for end in range(start + 1, len(s) + 1):
                if end in visited:
                    continue
                if s[start:end] in wordDict:
                    q.append(end)
                    visited.add(end)
        return False

File: leetcode/Medium/0139-word-break/0139-word-break.c
#define MAX_LENGTH 1000

bool wordBreak(char *s, char **wordDict, int wordDictSize) {
  int q[MAX_LENGTH];
  int front = 0, rear = 0;
  q[rear++] = 0;

  int visited[MAX_LENGTH];
  memset(visited, 0, sizeof(visited));

  while (front < rear) {
    int start = q[front++];
    if (start == strlen(s))
      return true;
    for (int end = start + 1; end <= strlen(s); ++end) {
      if (visited[end] == 1)
        continue;
      char word[MAX_LENGTH];
      strncpy(word, s + start, end - start);
      word[end - start] = '\0';

      for (int i = 0; i < wordDictSize; ++i) {
        if (strcmp(word, wordDict[i]) == 0) {
          q[rear++] = end;
          visited[end] = 1;
          break;
        }
      }
    }
  }

  return false;
}